### ISSUE TITLE
Project level attributes

### ISSUE THREAD

**Issue by musketyr** [asciidoctor/asciidoctor-intellij-plugin#740]
## Why the new feature should be added

In the current implementation there is an option to specify *Attributes* on a global level. It would be more useful to have this configuration also on the IntelliJ project's level because there is a lot of attributes which are bound to the current project. For example `root-dir` to refer to the root directory of the multi project. To see the preview in the Asciidoc editor I need to always change the value when I switch the projects.

## How the new feature should work
New extension point for project settings will be added holding the project-only attributes. These will be merged with the global attributes when rendering the preview in the editor.

see https://plugins.jetbrains.com/docs/intellij/settings-guide.html#declaring-project-settings

**Comment by ahus1** [comment]
Thanks for raising this issue. To support project specific attributes, there is a feature to add a special file `.asciidoctorconfig` with attributes. See https://intellij-asciidoc-plugin.ahus1.de/docs/users-guide/features/advanced/asciidoctorconfig-file.html for details.

There is also a special attribute `{asciidoctorconfigdir}` that allows you to specify a directory where the configuration.

Users can commit this file and make this configuration available for other IntelliJ users. The Eclipse plugin for AsciiDoc is planning a similar functionality in the future.

Please let me know if that works for you.

**Comment by musketyr** [comment]
Thanks, @ahus1!  It works great! But it's difficult to discover. It would be great if there is a quick fix available.

when `include file not found` violation happens then add another quick fix:
check if the include file URL contains a variable and try to replace it with the directory where the file is located
 - if the include can be resolved then create a file in the directory with `:variable-name: {asciidoctorconfigdir}`
 - if not then try the same for the project's root

**Comment by mojavelinux** [comment]
I like the idea of a quick fix.

Furthermore, I've noticed that in an Antora project, attributes defined in antora.yml override attributes in .asciidoctorconfig unless the attribute in antora.yml is soft set (the value ends with @). So the quick fix may need to suggest that.

To avoid having to change antora.yml, would it be possible to have it so that .asciidoctorconfig wins out regardless? After all, this configuration is only on the user's machine (or, if it is committed, Antora doesn't consider it). If you need me to file a separate issue, just say the word.

**Comment by mojavelinux** [comment]
I've also noticed that .asciidoctorconfig warns about attributes references for attributes that are defined in antora.yml, even though it works when the file is rendered in the preview.

**Comment by ahus1** [comment]
Hi @mojavelinux - thanks you for your interest in this issue. 

**TL;DR:** Soft-set attributes are a good rescue here. Another option could be an author's playbook (see end of this comment), or the plugins settings (which are never committed to the repo). In the current setup, the `.asciidoctorconfig` can never win as it is processed in the documents context.

Some historic context on `.asciidoctorconfig`: The goal was to find something that works both in the IDE and also on the command line with plain Asciidoctor / AsciiDoc. With https://github.com/asciidoctor/asciidoctor/issues/566 not being available, the common denominator is an include directive: While the IDE prepends it to every document in the folder in subfolders, you can choose to include it into your master document when running the rendering (for example in CI) on the command line. 

Having implemented this as an include, leaks some other features: When you add the extension `.adoc`, you can include other files from it, and can use conditionals like ifdef/ifndev/ifeval. It also leaks that it can't overwrite hard-set attributes. And it will always be last in picking up attributes. 

There are different places where the IDE picks up attribute values. The full overview in the order of precedence (see disclaimer below): 

1. Plugin settings (allows both soft and hard set attributes)
2. Antora playbooks (allows both soft and hard set attributes) 
3. Antora component descriptor (allows both soft and hard set attributes)
4. The prepended contents of `.asciidoctorconfig` (allows only setting attributes like as if it is running inside a document)

Disclaimer: While items one to three support soft and hard set attributes, there will definitely be corner cases that behave different from some people's expectations as they don't respect the hard/soft setting between each other: A "soft" plugin setting will overwrite a "hard" Antora playbook setting. This is probably an accidental part of implementation, still I didn't see bug raised for that, yet. If someone opens issues with reasons and examples to reproduce, that might change. 

Antora playbooks are identified by their name: as long as it ends with ".yml" and contains the words "antora" and "playbook" in the file name, it is considered an Antora playbook. If there are multiple files, their processing order for attributes is not guaranteed.

> would it be possible to have it so that .asciidoctorconfig wins out regardless?

Given all the above, and that is is processed in the document context, I'd say rather not.

> I've also noticed that .asciidoctorconfig warns about attributes references for attributes that are defined in antora.yml, even though it works when the file is rendered in the preview.

The attributes defined in a component descriptor are only valid within that component. So if you place an `.asciidoctorconfig` file outside of the component, it will show the attributes as unknown when editing it. When it is then prepended to an Antora page, those attributes are then known and it works as expected.

> To avoid having to change antora.yml, would it be possible to have it so that .asciidoctorconfig wins out regardless?

When I've been working with Antora projects, I usually have a playbook for the Author at hand to help with a locally pre-rendered site. I could imagine that the attribute setting you're trying to do here would be a match for that file (although I can also see possible conflicting interests, it depends on the setting).

**Comment by mojavelinux** [comment]
I understand your reasoning. I still think that the attributes in .asciidoctorconfig should win out regardless. It's especially useful when opening an Antora content root (the location of antora.yml) as a project directly. antora.yml cannot be changed and Antora doesn't permit alternate component version descriptors, so the overrides in .asciidoctorconfig become essential for configuring the environment for authoring. (The author playbook simply doesn't apply in this case because we're not working at the site level). However, I understand the tradeoff given the other functionality this file provides.

I do need to clarify something with regards to attribute resolution. You had said:

> The attributes defined in a component descriptor are only valid within that component. 

I'm talking about an .asciidoctorconfig that's adjacent to (in the same directory as) antora.yml. This situation arises when I open the folder that contains antora.yml as a project in IntelliJ. The attribute in question is defined in antora.yml, but when I try to reference that attribute in .asciidoctorconfig, the editor warns that the attribute is not defined. I would expect it to recognize it as a valid attribute and not warn. As I mentioned, it resolves in the preview, so this is just about the code assistance.

**Comment by ahus1** [comment]
I agree an override would be useful. Unfortunately the .asciidoctorconfig is running in the document's context and can't override a hard-set attribute. Going down a path making all attributes in the Antora component descriptor soft even when they aren't will IMHO lead to other surprises. So for now, I suggest adding the "@" to the Antora component descriptor.

I did a small test when it warns about a non-defined attribute, and there is a bit more detail about it: 

1. When it is placed in the same folder as `antora.yml`, I see the warning as you describe
2. When it is placed in the folder `modules` below `antora,yml`, the warning vanishes, but the preview doesn't show the contents of the attribute yet
3. When it is placed in a folder below `modules` for example `ROOT` the preview shows the contents of the variable. This is also the first place where the Antora stylesheet is picked up

1+2 are somewhat of a grey area, as there are usually no pages in an Antora structure in those folders; so I haven't tested this before. 3 is then the first place where for example navigation files show up for the first time, so attributes and Antora style sheets should be picked up. 

If you want to keep the file in the parent folder, and suppress the warning, you can add the following line at the first line of the file: 

```
// suppress inspection "AsciiDocAttributeShouldBeDefined" for whole file
```

> Antora doesn't permit alternate component version descriptors 

Maybe you can add more context or an example to this use case (possibly in a new issue), so we can discuss alternative ideas to satisfy that need.
